cacheproc/cacheproc.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2023 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package cacheproc implements the mechanics of talking to cmd/go's GOCACHE protocol
+// so you can write a caching child process at a higher level.
+package cacheproc
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"sync"
+
+	"github.com/bradfitz/go-tool-cache/cachers"
+
+	"github.com/bradfitz/go-tool-cache/wire"
+)
+
+// Process implements the cmd/go JSON protocol over stdin & stdout via three
+// funcs that callers can optionally implement.
+type Process struct {
+	cache cachers.LocalCache
+}
+
+func NewCacheProc(cache cachers.LocalCache) *Process {
+	return &Process{
+		cache: cache,
+	}
+}
+
+func (p *Process) Run() error {
+	br := bufio.NewReader(os.Stdin)
+	jd := json.NewDecoder(br)
+
+	bw := bufio.NewWriter(os.Stdout)
+	je := json.NewEncoder(bw)
+	if err := p.cache.Start(); err != nil {
+		return err
+	}
+	caps := []wire.Cmd{"get", "put", "close"}
+	je.Encode(&wire.Response{KnownCommands: caps})
+	if err := bw.Flush(); err != nil {
+		return err
+	}
+
+	var wmu sync.Mutex // guards writing responses
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	for {
+		var req wire.Request
+		if err := jd.Decode(&req); err != nil {
+			if errors.Is(err, io.EOF) {
+				return nil
+			}
+			return err
+		}
+		if req.Command == wire.CmdPut && req.BodySize > 0 {
+			// TODO(bradfitz): stream this and pass a checksum-validating
+			// io.Reader that validates on EOF.
+			var bodyb []byte
+			if err := jd.Decode(&bodyb); err != nil {
+				log.Fatal(err)
+			}
+			if int64(len(bodyb)) != req.BodySize {
+				log.Fatalf("only got %d bytes of declared %d", len(bodyb), req.BodySize)
+			}
+			req.Body = bytes.NewReader(bodyb)
+		}
+		go func() {
+			res := &wire.Response{ID: req.ID}
+			ctx := ctx // TODO: include req ID as a context.Value for tracing?
+			if err := p.handleRequest(ctx, &req, res); err != nil {
+				res.Err = err.Error()
+			}
+			wmu.Lock()
+			defer wmu.Unlock()
+			je.Encode(res)
+			bw.Flush()
+		}()
+	}
+}
+
+func (p *Process) handleRequest(ctx context.Context, req *wire.Request, res *wire.Response) error {
+	switch req.Command {
+	default:
+		return errors.New("unknown command")
```

Comment:
NIT: make it a global: `var ErrUnknownCommand = errors.New("unknown command")` (this way you can later `errors.Is(err, ErrUnknownCommand)`)
```suggestion
		return ErrUnknownCommand
```